httpcore5/src/main/java/org/apache/hc/core5/http/impl/io/DefaultResponseOutOfOrderStrategy.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.core5.http.impl.io;
+
+import org.apache.hc.core5.annotation.Contract;
+import org.apache.hc.core5.annotation.ThreadingBehavior;
+import org.apache.hc.core5.http.ClassicHttpRequest;
+import org.apache.hc.core5.http.io.HttpClientConnection;
+import org.apache.hc.core5.http.io.ResponseOutOfOrderStrategy;
+import org.apache.hc.core5.util.Timeout;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Default {@link ResponseOutOfOrderStrategy} implementation which checks for premature responses every 8 KiB based
+ * on testing using values between 4 KiB and 128 KiB. This is optimized for correctness and results in a maximum
+ * upload speed of 8 MiB/s.
+ *
+ * @since 5.1
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE)
+public final class DefaultResponseOutOfOrderStrategy implements ResponseOutOfOrderStrategy {
+
+    public static final DefaultResponseOutOfOrderStrategy INSTANCE = new DefaultResponseOutOfOrderStrategy();
+
+    private static final int CHUNK_SIZE = 8 * 1024;
+
+    @Override
+    public boolean checkForEarlyResponse(
```

Comment:
I think the name is a bit misleading. Since it is a boolean and used as such with an exception, shouldn't it be `isEarlyResponse` or similar?

```java
if (isEarlyResponse())
  throw new ...Exception();
```

Reply:
> There are only two hard things in Computer Science: cache invalidation and naming things.
> -- Phil Karlton

Perhaps `isEarlyResponseAvailable` or `isEarlyResponseDetected`? I think I prefer `isEarlyResponseDetected` because the no-op implementation may not detect an early response even if one is available.

Reply:
Nice citation. I pretty much like the new change, happily accepting.